lean_source/06_Abstract_Algebra/source_01_Structures.lean

```diff
@@ -0,0 +1,503 @@
+putting it together again in different ways.
+It therefore makes sense that Lean and mathlib offer so many ways
+of doing this efficiently.
+As an exercise, try proving that ``point.add`` is associative.
+Then define scalar multiplication for a point and show that it
+distributes over addition.
+You can use pattern matching or projections in the definition,
+as you prefer.
+BOTH: -/
+-- QUOTE:
+theorem add_assoc (a b c : point) : (a.add b).add c = a.add (b.add c) :=
+/- EXAMPLES:
+sorry
+SOLUTIONS: -/
+by { cases a, cases b, cases c, simp [add, add_assoc] }
+-- BOTH:
+
+/- EXAMPLES:
+def smul (r : ℝ) : point → point := sorry
+SOLUTIONS: -/
+def smul (r : ℝ) : point → point
+| ⟨x, y, z⟩ := ⟨r * x, r * y, r * z⟩
+-- BOTH:
+
+theorem smul_distrib (r : ℝ) (a b : point) :
+  (smul r a).add (smul r b) = smul r (a.add b) :=
+/- EXAMPLES:
+sorry
+SOLUTIONS: -/
+by { cases a, cases b, simp [add, smul, mul_add] }
+-- BOTH:
+-- QUOTE.
+
+end point
+
+/- TEXT:
+Using structures is only the first step on the road to
+algebraic abstraction.
+We don't yet have a way to link ``point.add`` to the generic ``+`` symbol,
+or to connect ``point.add_comm`` and ``point.add_assoc`` to
+the generic ``add_comm`` and ``add_assoc`` theorems.
+These tasks belong to the *algebraic* aspect of using structures,
+and we will explain how to do carry them out in the next section.
+For now, just think of a structure as a way of bundling together objects
+and information.
+
+It is especially useful that a structure can specify not only
+data types but also constraints that the data must satisfy.
+In Lean, the latter are represented as fields of type ``Prop``.
+For example, the *standard 2-simplex* is defined to be the set of
+points :math:`(x, y, z)` satisfying :math:`x ≥ 0`, :math:`y ≥ 0`, :math:`z ≥ 0`,
+and :math:`x + y + z = 1`.
+If you are not familiar with the notion, you should draw a picture,
+and convince yourself that this set is
+the equilateral triangle in three-space with vertices
+:math:`(1, 0, 0)`, :math:`(0, 1, 0)`, and :math:`(0, 0, 1)`,
+together with its interior.
+We can represent it in Lean as follows:
+BOTH: -/
+-- QUOTE:
+structure standard_two_simplex :=
+(x : ℝ)
+(y : ℝ)
+(z : ℝ)
+(x_nonneg : 0 ≤ x)
+(y_nonneg : 0 ≤ y)
+(z_nonneg : 0 ≤ z)
+(sum_eq   : x + y + z = 1)
+-- QUOTE.
+
+/- TEXT:
+Notice that the last four fields refer to ``x``, ``y``, and ``z``,
+that is, the first three fields.
+We can define a map from the two-simplex to itself that swaps ``x`` and ``y``:
+BOTH: -/
+namespace standard_two_simplex
+
+-- EXAMPLES:
+-- QUOTE:
+def swap_xy : standard_two_simplex → standard_two_simplex
+| ⟨x, y, z, xnn, ynn, znn, hsum⟩ :=
+  { x := y,
+    y := x,
+    z := z,
+    x_nonneg := ynn,
+    y_nonneg := xnn,
+    z_nonneg := znn,
+    sum_eq   := by rwa add_comm y x }
+-- QUOTE.
+
+/- TEXT:
+More interestingly, we can compute the midpoint of two points on
+the simplex. We need to add ``noncomputable theory`` in order to
+use division on the real numbers.
```

Comment:
Do we have any TODO file where we would register something like:
```quote
Once we have a somewhat in-depth explanation of `noncomputable`, go back to the simplex example in the structure chapter and add a link.
```
Independently from this question, should we put links to TPIL as placeholder?

Reply:
There is an `OMIT` directive in my markup that we can use to leave notes to ourselves in the files that doesn't get included in any output. I added this one with a label `TODO`. It's easy to search in VS Code for all the `TODO` labels.